components/sync15/src/sync_multiple.rs

```diff
@@ -102,160 +130,301 @@ pub fn sync_multiple(
-                    log::error!(
-                        "Failed to parse PersistedGlobalState from JSON! Falling back to default"
-                    );
-                    PersistedGlobalState::default()
-                }
-            }
+        log::info!("Entering sync state machine");
+        // Advance the state machine to the point where it can perform a full
+        // sync. This may involve uploading meta/global, crypto/keys etc.
+        let mut global_state = self.run_state_machine(&client_info, &mut pgs)?;
+
+        if self.was_interrupted() {
+            return Ok(());
         }
-        None => {
-            log::info!("The application didn't give us persisted state - this is only expected on the very first run for a given user.");
-            PersistedGlobalState::default()
+
+        // Set the service status to OK here - we may adjust it based on an individual
+        // store failing.
+        self.result.service_status = ServiceStatus::Ok;
+
+        log::info!("Synchronizing stores");
+
+        let telem_sync = self.sync_stores(&client_info, &mut global_state);
+        self.result.telemetry.sync(telem_sync);
+
+        log::info!(
+            "Finished syncing stores. All successful: {}",
+            !self.any_failed_engines
+        );
+
+        if self.any_failed_engines {
+            // XXX - not clear if we should really only do this on full success,
+            // particularly if it's just a network error. See XXX above for more.
+            log::info!("Updating persisted global state");
+            self.mem_cached_state.last_client_info = Some(client_info);
+            self.mem_cached_state.last_global_state = Some(global_state);
         }
-    };
 
-    if interruptee.was_interrupted() {
-        sync_result.service_status = ServiceStatus::Interrupted;
-        return Ok(());
+        Ok(())
+    }
+
+    fn was_interrupted(&mut self) -> bool {
+        if self.interruptee.was_interrupted() {
+            log::info!("Interrupted, bailing out");
+            self.result.service_status = ServiceStatus::Interrupted;
+            true
+        } else {
+            false
+        }
     }
 
-    // Advance the state machine to the point where it can perform a full
-    // sync. This may involve uploading meta/global, crypto/keys etc.
-    let global_state = {
-        let last_state = mem::replace(&mut mem_cached_state.last_global_state, None);
+    fn sync_stores(
+        &mut self,
+        client_info: &ClientInfo,
+        global_state: &mut GlobalState,
+    ) -> telemetry::SyncTelemetry {
+        let mut telem_sync = telemetry::SyncTelemetry::new();
+        for store in self.stores {
+            let name = store.collection_name();
+            if self
+                .backoff
+                .get_required_wait(self.ignore_soft_backoff)
+                .is_some()
+            {
+                log::warn!("Got backoff, bailing out of sync early");
```

Comment:
Hmm, where do we set `ServiceStatus::BackedOff`?

**Edit**: Oh, where we do `ServiceStatus::from_err` below? What if we've been told to back off before...do we need to set it here?